docs/abi.rst

```diff
@@ -0,0 +1,919 @@
+.. _abi:
+
+ABI Support
+===========
+
+.. warning::
+    ABI support is still taking shape and is subject to backwards incompatible changes.
+        * Based on feedback, the API and usage patterns are likely to change.
+        * For the following use cases, feel encouraged to rely on abstractions. Expect a best-effort attempt to minimize backwards incompatible changes along with a migration path.
+
+            * :any:`ABIReturnSubroutine` usage for ABI Application entry point definition.
+
+            * :any:`Router` usage for defining how to route program invocations.
+        * For general purpose :any:`Subroutine` definition usage, use at your own risk. Based on feedback, the API and usage patterns will change more freely and with less effort to provide migration paths.
+
+    For these reasons, we strongly recommend using :any:`pragma` or the :any:`Pragma` expression to pin the version of PyTeal in your source code.
```

Comment:
@jasonpaulos Sanity checking:  Is it by-design that the user guide does _not_ provide pragma explanation beyond the API docs?

* I'm asking because we merged https://github.com/algorand/pyteal/pull/429 with the intention to revisit user guide docs here.  And I'm trying to gauge opinion on the topic.
* I'm open to being convinced otherwise, but my starting position is:  The concept is sufficiently important to warrant a 1st class user guide mention.
  * Also flexible:  I can imagine extending https://pyteal.readthedocs.io/en/stable/versions.html or reworking https://pyteal.readthedocs.io/en/stable/compiler_optimization.html to include optimizations + version pragma.

Reply:
Added documentation to the versions page, please re-open if anything is missing